File: app/src/UI/Features/Records/Activity/forms/common/validators/distinctEntries.ts
/**
 * @desc Checks for uniqueness based on a specific set of keys.
 * @param arr The array of objects to validate.
 * @param uniqueKeys The specific keys that, when combined, must be unique.
 * @param errorMessage Human-readable name for the error message.
 */
const distinctEntries = <T extends Record<string, unknown>>(
  arr: T[],
  uniqueKeys: (keyof T)[],
  errorMessage: string
): boolean | string => {
  const seen = new Set<string>();
  for (const entry of arr) {
    // Build out unique strings based on unique keys.
    const compositeId = uniqueKeys
      .map((k) => {
        const val = entry[k];
        // Normalize values to strings, handling nulls/undefined
        return val !== null && val !== undefined ? String(val).trim() : '';
      })
      .join('-');
    // Don't validate on Empty entries.
    if (compositeId.replace(/-/g, '') === '') continue;
    if (seen.has(compositeId)) {
      return errorMessage;
    }
    seen.add(compositeId);
  }
  return true;
};

export default distinctEntries;

File: api-rework/invasives/api/protocol/activity/validators/distinct_entries.py
from typing import Any, Dict, List, Union


def distinct_entries(
    arr: List[Dict[str, Any]], unique_keys: List[str], error_message: str
) -> Union[bool, str]:
    """
    Checks for uniqueness based on a specific set of keys.

    :param arr: The list of dictionaries to validate.
    :param unique_keys: The specific keys that, when combined, must be
    unique.
    :param error_message: Human-readable name for the error message.
    :return: True if all non-empty entries are unique; error_message otherwise.
    """
    seen: set[str] = set()

    for entry in arr:
        # Build composite string based on unique keys
        parts: list[str] = []
        for key in unique_keys:
            val = entry.get(key) if isinstance(entry, dict) else None
            # Normalize values to strings, handling None
            parts.append(str(val).strip() if val is not None else "")

        composite_id = "-".join(parts)

        # Skip validation on empty entries
        if composite_id.replace("-", "") == "":
            continue

        if composite_id in seen:
            return error_message

        seen.add(composite_id)

    return True
